[doc] Re-update documentation on autodiff (#1623)
taichi-dev/taichi: docs/autodiff.rst

.. _differentiable:

Differentiable programming
==========================

We suggest starting with the ``ti.Tape()``, and then migrate to more advanced differentiable programming using the ``kernel.grad()`` syntax if necessary.

Introduction
------------

For example, you have the following kernel:

.. code-block:: python

    x = ti.var(ti.f32, ())
    y = ti.var(ti.f32, ())

    @ti.kernel
    def compute_y():
        y[None] = ti.sin(x[None])

> Now you want to get the derivative of y corresponding to x, i.e. dy/dx.
> You may want to write the derivatived kernel by hand:

Reviewer: ```suggestion
Now if you want to get the derivative of y corresponding to x, i.e., dy/dx.
You may want to implement the derivative kernel by yourself:
```
There should be a comma behind i.e. right? It's the abbreviation. And using the adj of derivative should be enough? wdyt :)

.. code-block:: python

    x = ti.var(ti.f32, ())
    y = ti.var(ti.f32, ())
    dy_dx = ti.var(ti.f32, ())

    @ti.kernel
    def compute_dy_dx():
        dy_dx[None] = ti.cos(x[None])

But wait, what if I changed the original ``compute_y``? We will have to
recalculate the derivative by hand and rewrite ``compute_dy_dx`` again, which

> is very error-prone and not convienent at all.

Reviewer: ```suggestion
is very error-prone and not convenient at all.
```
typo

> If you meet this situation, don't worry! Taichi provides a handy autodiff

Reviewer: ```suggestion
If this situation occurs, don't worry! Taichi provides a handy autodiff
```

system that can help you obtain the derivative of a kernel without any pain!

Using ``ti.Tape()``
-------------------

Let's still take the ``compute_y`` in above example for explaination.
What's the most convienent way to obtain a kernel that computes x to dy/dx?

1. Use the ``needs_grad=True`` option when declaring tensors involved in the
   derivative chain.

> 2. Use ``with ti.Tape(y):`` to enbrace the invocation into kernel(s) you want
>    to compute derivative.

Reviewer: IIUIC this means we use `with.ti.Tape(y)` to allow the invocation into kernels? 🤔 

```suggestion
2. Use ``with ti.Tape(y):`` to embrace the invocation into kernel(s) you want
   to compute derivative.
```

3. Now ``x.grad[None]`` is the dy/dx value at current x.

.. code-block:: python

    x = ti.var(ti.f32, (), needs_grad=True)
    y = ti.var(ti.f32, (), needs_grad=True)

    @ti.kernel
    def compute_y():
        y[None] = ti.sin(x[None])

    with ti.Tape(y):
        compute_y()

    print('dy/dx =', x.grad[None])
    print('at x =', x[None])

It's equivalant to:

.. code-block:: python

    x = ti.var(ti.f32, ())
    y = ti.var(ti.f32, ())
    dy_dx = ti.var(ti.f32, ())

    @ti.kernel
    def compute_dy_dx():
        dy_dx[None] = ti.cos(x[None])

    compute_dy_dx()

    print('dy/dx =', dy_dx[None])
    print('at x =', x[None])

Usage example
+++++++++++++

For a physical simulation, sometimes it could be easy to compute the energy but
hard to compute the force on each particles.

But recall that we can differentiate (negative) potential energy to get forces.
a.k.a.: ``F_i = -dU / dx_i``.

> So once you've write a kernel that is able to compute the potential energy,
> then you may use Taichi's autodiff system to obtain the derivative of it and
> therefore obtain the force on each particles.

Reviewer: ```suggestion
So once you've write a kernel that is able to compute the potential energy,
you may use Taichi's autodiff system to obtain the derivative of it and
then the force on each particles.
```
